tree_test.go

```diff
@@ -483,6 +489,163 @@ func TestLeafCounts(t *testing.T) {
+	leaf4Hash := leafSum(hash, leaf4Data)
+	err1 := tree2.PushSubTree(0, leaf1Hash)
+	err2 := tree2.PushSubTree(0, leaf2Hash)
+	err3 := tree2.PushSubTree(0, leaf3Hash)
+	err4 := tree2.PushSubTree(0, leaf4Hash)
+	if err := errors.Compose(err1, err2, err3, err4); err != nil {
+		t.Fatal(err)
+	}
+	if !bytes.Equal(tree2.Root(), expectedRoot) {
+		t.Fatal("root doesn't match expected root")
+	}
+
+	// Create 2 height 1 subtrees and combine them. The root should be the
+	// same.
+	tree3 := New(hash)
+	node12Hash := nodeSum(hash, leaf1Hash, leaf2Hash)
+	node34Hash := nodeSum(hash, leaf3Hash, leaf4Hash)
+	err1 = tree3.PushSubTree(1, node12Hash)
+	err2 = tree3.PushSubTree(1, node34Hash)
+	if err := errors.Compose(err1, err2); err != nil {
+		t.Fatal(err)
+	}
+	if !bytes.Equal(tree3.Root(), expectedRoot) {
+		t.Fatal("root doesn't match expected root")
+	}
+
+	// Create 1 height 2 subtree and add it to the tree. The root should be the
+	// same.
+	tree4 := New(hash)
+	node1234Hash := nodeSum(hash, node12Hash, node34Hash)
+	if err := tree4.PushSubTree(2, node1234Hash); err != nil {
+		t.Fatal(err)
+	}
+	if !bytes.Equal(tree4.Root(), expectedRoot) {
+		t.Fatal("root doesn't match expected root")
+	}
+
+	// Create 1 height 1 tree and add 2 height 0 trees. The root should be the
+	// same.
+	tree5 := New(hash)
+	err1 = tree5.PushSubTree(1, node12Hash)
+	err2 = tree5.PushSubTree(0, leaf3Hash)
+	err3 = tree5.PushSubTree(0, leaf4Hash)
+	if err := errors.Compose(err1, err2, err3); err != nil {
+		t.Fatal(err)
+	}
+	if !bytes.Equal(tree5.Root(), expectedRoot) {
+		t.Fatal("root doesn't match expected root")
+	}
+
+	// Create 1 height 1 tree and add 2 leaves. The root should be the same.
+	tree6 := New(hash)
+	if err := tree6.PushSubTree(1, node12Hash); err != nil {
+		t.Fatal(err)
+	}
+	tree6.Push(leaf3Data)
+	tree6.Push(leaf4Data)
+	if !bytes.Equal(tree6.Root(), expectedRoot) {
+		t.Fatal("root doesn't match expected root")
+	}
+}
+
+// TestPushSubTreeSimple tests pushing some valid and invalid subTrees to the
+// tree.
+func TestPushSubTreeSimple(t *testing.T) {
+	tree := New(sha256.New())
+
+	// Add a subTree of height 5 to the empty tree.
+	if err := tree.PushSubTree(5, []byte{}); err != nil {
+		t.Fatal(err)
+	}
+	if tree.Root() == nil {
+		t.Fatal("root should not be nil after adding a subTree")
+	}
+	// Add a subTree of a height >5 to the tree. This should not be possible.
+	if err := tree.PushSubTree(6, []byte{}); err == nil {
+		t.Fatal("pushing a subTree with a larger height than the smallest subTree should fail")
+	}
+	// The current index should be 2^5
+	expectedIndex := uint64(1 << 5)
+	if tree.currentIndex != expectedIndex {
+		t.Errorf("expected index %v but was %v", expectedIndex, tree.currentIndex)
+	}
+	// Add a subTree of the same height as the smallest subTree in the merkle
+	// tree and check again.
+	if err := tree.PushSubTree(5, []byte{}); err != nil {
+		t.Fatal(err)
+	}
+	if tree.currentIndex != 2*expectedIndex {
```

Comment:
this is confusing to read, should probably double the expected index value after you push, and then compare to the updated value.